Docs/doc 95/doc 105/technical term expectation suite (#4105)
great-expectations/great_expectations: docs/terms/expectation_suite.md

title: Expectation Suite
hoverText: A collection of verifiable assertions about data.
---

import UniversalMap from '/docs/images/universal_map/_universal_map.mdx';
import TechnicalTag from '../term_tags/_tag.mdx';
import CreateHeader from '/docs/images/universal_map/_um_create_header.mdx';
import ValidateHeader from '/docs/images/universal_map/_um_validate_header.mdx';

<UniversalMap setup='inactive' connect='inactive' create='active' validate='active'/> 

## Overview

### Definition

An Expectation Suite is a collection of verifiable assertions about data.

### Features and promises

> Expectation Suites combine multiple <TechnicalTag relative="../" tag="expectation" text="Expectations" /> into an overall description of a dataset. For example, a team can group all the Expectations about a given table in given database into an Expectation Suite and call it `my_database.my_table`. Note these names are completely flexible and the only constraint on the name of a suite is that it must be unique to a given project.

Reviewer: To avoid implicitly bringing in the question of what a "dataset" is I think it makes sense to either just say "data" (recommendation here) or to go ahead and describe this as a description of a Data Asset.

My only hesitation with the latter approach is that an Expectation Suite can be applied to multiple data assets, and often is used that way.

```suggestion
Expectation Suites combine multiple <TechnicalTag relative="../" tag="expectation" text="Expectations" /> into an overall description of data. For example, a team can group all the Expectations about a given table in given database into an Expectation Suite and call it `my_database.my_table`. Note these names are completely flexible and the only constraint on the name of a suite is that it must be unique to a given project.
```

### Relationship to other objects

> Expectation Suites are stored in an <TechnicalTag relative="../" tag="expectation_store" text="Expectation Store" />.  They are generated by <TechnicalTag relative="../" tag="profiler" text="Profilers" />, and are used by <TechnicalTag relative="../" tag="checkpoint" text="Checkpoints" /> to <TechnicalTag relative="../" tag="validation" text="Validate" /> data.

Reviewer: ```suggestion
Expectation Suites are stored in an <TechnicalTag relative="../" tag="expectation_store" text="Expectation Store" />.  They are generated interactively using a <TechnicalTag relative="../" tag="validator" text="Validator" /> or automatically using <TechnicalTag relative="../" tag="profiler" text="Profilers" />, and are used by <TechnicalTag relative="../" tag="checkpoint" text="Checkpoints" /> to <TechnicalTag relative="../" tag="validation" text="Validate" /> data.
```

(I'm not sure "validator" is already in the glossary, and I'm open to pointing to another ref, but wanted to highlight that most users today are still manually building expectations)

Author: Doesn't the interactive workflow start with getting a Validator from a Data Context?  I'm referencing https://docs.greatexpectations.io/docs/guides/expectations/how_to_create_and_edit_expectations_with_instant_feedback_from_a_sample_batch_of_data and the Jupyter notebook generated in this process.  All of the expectations are created with some variation of `validator.expect_something` which was why I went with this explanation.  I was told that writing Expectations based purely on domain knowledge (ie, manually writing them as config statements with no reference to source data) was considered a workflow for advanced users and not the recommended process.

For more information on these processes, please see:
- [Our overview on the process of Creating Expectations](../guides/expectations/create_expectations_overview.md)
- [Our guide on how to create and edit Expectations with a Profiler](../guides/expectations/how_to_create_and_edit_expectations_with_a_profiler.md)
- [Our guide on how to create and edit Expectations with instant feedback from a sample Batch of data](../guides/expectations/how_to_create_and_edit_expectations_with_instant_feedback_from_a_sample_batch_of_data.md)

<ValidateHeader/>

Expectation Suites are *used* during the Validation of data.  In this step, you will need to provide one or more Expectation Suites to a Checkpoint.  This can either be done by configuring the Checkpoint to use a preset list of one or more Expectation Suites, or by configuring the Checkpoint to accept a list of one or more Expectation Suites at runtime.

## Features

### CRUD operations

A Great Expectations Expectation Suite enables you to perform Create, Read, Update, and Delete (CRUD) operations on the Suite's Expectations without needing to re-run them. 

### Reusability

> Expectation Suites are primarily used by Checkpoints, which can accept a list of one or more Expectation Suite.  Because they are stored independently of the Checkpoints that use them, the same Expectation Suite can be included in the list for multiple Checkpoints, provided the Expectation Suite contains a list of Expectations that describe the data that Checkpoint will Validate.  This allows you to, for example, create multiple Checkpoints with increasing vigor of in their validation, or create multiple Checkpoints with the same base set of Expectations and some additional Expectations tailored to specific data (such as one Checkpoint for validating raw data, and one with additional Expectation Suites for validating post-processing data).

Reviewer: This section is confusing to me. I understand the notion that the same suite can be applied to multiple datasets, but I'm not sure where we're trying to go with the discussion of increasing rigor. Can you clarify?

Author: I'm basing this on my experience working in GIS.  We would often be given raw data that would then need to be formatted using automated processes before being passed on to technicians to have additional details filled in and records verified against alternate datasources.  Ultimately, we would end up with at least three related levels of validation being needed, covering data that had increasing rigidity of its standardization -- raw data, pre-(manual) processing, and post (manual) processing.  The most sensible way of doing this in GE, I thought, would be to start with one Expectation Suite and Checkpoint for the raw data, to detect changes that may have occurred there.  A second Checkpoint would include that first Expectation Suite and an additional Expectation Suite to validate that none of the automated steps in generating the pre-(manual) processing data had behaved unexpectedly, as can happen when scripts are being refined (as would be required if the first Checkpoint flagged a change in how the raw data was formatted) while production is underway.  And then a third Checkpoint which combined the first two Expectation Suites and a third Expectation Suite asserting that all required fields were populated, etc, would be applied as quality control to the post-process data.  With each step in the cleaning, clarifying, and expanding of the data you would ensure that the previous Expectations hadn't been violated as a sort of quality control check, while also verifying that all of the additional Expectations you would have for data that had gone through an automated cleanup and then through technician processing were also being met, without having to rewrite the Expectations from the earlier iterations of the data.

Reviewer: Discussed live; @Rachel-Reverie is verifying the behavior here, but we think it's the case that the list of validations is a list of (batch_request, suite) pairs.

Author: Verified: Checkpoints accept one or more Execution Suite and Batch Request pairs.  I've updated this section accordingly.

## API basics

### CRUD operations

Each of the Expectation Suite methods that support a Create, Read, Update, or Delete (CRUD) operation relies on two main parameters - `expectation_configuration` and `match_type`.

- **expectation_configuration** - an `ExpectationConfiguration` object that is used to determine whether and where this Expectation already exists within the Suite. It can be a complete or a partial ExpectationConfiguration.
- **match_type** - a string with the value of `domain`, `success`, or `runtime` which determines the criteria used for matching:

>     - `domain` checks whether two Expectation Configurations apply to the same data. It results in the loosest match, and can use the least complete ExpectationConfiguration object. For example, for a column map Expectation, `domain_kwargs` will include the Expectation_type, the column, and any row_conditions that may affect which rows are evaluated by the Expectation.

Reviewer: ```suggestion
    - `domain` checks whether two Expectation Configurations apply to the same data. It results in the loosest match, and can use the least complete ExpectationConfiguration object. For example, for a column map Expectation, a `domain` **match_type** will check that the expectation_type matches, and that the column and any row_conditions that affect which rows are evaluated by the Expectation match.
```

    - `success` criteria are more exacting - in addition to the `domain` kwargs, these include those kwargs used when evaluating the success of an Expectation, like `mostly`, `max`, or `value_set`.
    -`runtime` are the most specific - in addition to `domain_kwargs` and `success_kwargs`, these include kwargs used for runtime configuration. Currently, these include `result_format`, `include_config`, and `catch_exceptions`

### How to access

You will rarely need to directly access an Expectation Suite.  If you do need to edit one, the simplest way is through the CLI.  To do so, run the command:

```markdown title="Terminal command"

> great_expectations suite edit NAME_OF_YOUR_SUITE_HERE

Reviewer: I think we should mention the choice of whether to load data to edit the suite or not; I think there's a specific flag for that.

Author: I'll look into that.

Author: It looks like there is a `-ds, --datasource-name TEXT` flag.  is that the one you are thinking of?  Or the `-br, -batch-request TEXT` option?  (Perhaps both?)  Both are mutually exclusive and assume the use of the `--interactive` flag.

Reviewer: I think the flag I was thinking of is the `--interactive` one...but I was thinking that was default. If that's not the cause (i.e. the default is to NOT apply interactive) then I think this is already correct as written.

Author: The default is interactive, but if you don't specify the CLI will prompt you to ask if you want to edit interactively or manually.
